
package fr.proto.fs;

option cc_generic_services = true;

message handle {
    optional uint32 value = 1;
}

message file_position {
    optional uint64 position = 1;
}

message empty { }

// fs directory operations;
message handle_path {
    optional handle  hdl    = 1;
    optional string  path   = 2;
}

message element_stat {  /* man 2 stat */
    optional uint64  dev     =  1; /* ID of device containing file */
    optional uint64  ino     =  2; /* inode number */
    optional uint32  mode    =  3; /* protection */
    optional uint64  nlink   =  4; /* number of hard links */
    optional uint32  uid     =  5; /* user ID of owner */
    optional uint32  gid     =  6; /* group ID of owner */
    optional uint64  rdev    =  7; /* device ID (if special file) */
    optional uint64  size    =  8; /* total size, in bytes */
    optional uint32  blksize =  9; /* blocksize for filesystem I/O */
    optional uint64  blocks  = 10; /* number of 512B blocks allocated */
    optional uint64  atime   = 11; /* time of last access */
    optional uint64  mtime   = 12; /* time of last modification */
    optional uint64  ctime   = 13; /* time of last status change */
}

message element_info {
    optional bool is_exist      = 1;
    optional bool is_directory  = 2;
    optional bool is_empty      = 3;
    optional bool is_regular    = 4;
    optional bool is_symlink    = 5;
}

message iterator_info {
    optional handle  hdl  = 1;
    optional string  path = 2; // path
    optional bool    end  = 3; // this is the last one
}

message rename_req {
    optional handle_path  src = 1;
    optional handle_path  dst = 2;
}

message read_file_req {
    optional handle_path  dst = 1;
    optional uint32       len = 2;
}

message read_file_res {
    optional bytes data = 1;
}

message write_file_req {
    optional handle_path  dst  = 1;
    optional bytes        data = 2;
}

service instance {

    // open instance
    rpc open( handle_path )        returns ( handle_path );

    // change instance path
    rpc cd  ( handle_path )        returns ( handle_path );

    // get path by handle
    rpc pwd ( handle_path )        returns ( handle_path );

    // check if path exists
    rpc exists( handle_path )      returns ( element_info );

    // get file size
    rpc file_size( handle_path )   returns ( file_position );

    // get path info
    rpc info( handle_path )        returns ( element_info );

    // get path stat
    rpc get_stat( handle_path )    returns ( element_stat );

    // make directory
    rpc mkdir( handle_path )       returns ( handle_path );

    // rename file element
    rpc rename( rename_req )       returns ( handle_path );

    // del fs element
    rpc del( handle_path )         returns ( handle_path );

    // del fs element or fs tree
    rpc delall( handle_path )      returns ( handle_path );

    // iterate directory
    rpc iter_begin(  handle_path  ) returns( iterator_info );
    rpc iter_next ( iterator_info ) returns( iterator_info );
    rpc iter_info ( iterator_info ) returns( element_info  );
    rpc iter_clone( iterator_info ) returns( iterator_info );

    // close instance or iterator;
    rpc close( handle ) returns ( empty );

    /// read and write file fast

    rpc read_file( read_file_req )   returns( read_file_res );
    rpc write_file( write_file_req ) returns( empty );

}

message open_file_values { /* man 2 open */
    enum flags {
        RDONLY      =        0;
        WRONLY      =       01;
        RDWR        =       02;
        CREAT       =     0100;
        EXCL        =     0200;
        TRUNC       =    01000;
        APPEND      =    02000;
        NONBLOCK    =    04000;
        ASYNC       =   020000;
        SYNC        = 04010000;
    }

    enum mode {
        IRWXU  = 00700;
        IRUSR  = 00400;
        IWUSR  = 00200;
        IXUSR  = 00100;
        IRWXG  = 00070;
        IRGRP  = 00040;
        IWGRP  = 00020;
        IXGRP  = 00010;
        IRWXO  = 00007;
        IROTH  = 00004;
        IWOTH  = 00002;
        IXOTH  = 00001;
    }
}

message file_open_req {
    optional string path  = 1;
    optional uint32 flags = 2; // see open_file_values.flags
    optional uint32 mode  = 3; // see open_file_values.mode
}

enum file_position_whence {
    POS_SEEK_SET  = 0;
    POS_SEEK_CUR  = 1;
    POS_SEEK_END  = 2;
}

message file_set_position {
    optional handle  hdl      = 1;
    optional int64   position = 2;
    optional uint32  whence   = 3;
}

message file_data_block {
    optional handle  hdl     = 1;
    optional bytes   data    = 2;
    optional uint32  length  = 3;
}

message register_req {
    optional handle  hdl = 1;
}

message register_res {
    optional uint32  async_op_id  = 1;
}

// files operations
service file {

    rpc open(  file_open_req ) returns ( handle );

    rpc tell( handle)             returns ( file_position );
    rpc seek( file_set_position ) returns ( file_position );

    rpc read(  file_data_block )  returns ( file_data_block );
    rpc write( file_data_block )  returns ( file_data_block );

    rpc flush( handle )  returns ( empty );

    rpc register_for_events( register_req ) returns ( register_res );
    rpc unregister ( register_req ) returns ( empty );

    rpc close( handle)   returns ( empty );

}

